File: GoInterviewQuestion/shopee/find/34. 在排序数组中查找元素的第一个和最后一个位置.go
package find


func SearchRange(nums []int, target int) []int {
	length := len(nums)
	if length == 0 || target > nums[length - 1] || target < nums[0] {
		return []int{-1, -1}
	}

	ret := make([]int, 0)
	i, j := 0, length - 1
	//找到最左边出现的位置
	for i < j {
		mid := (i+j)/2
		if target <= nums[mid] {
			j = mid
		} else {
			i = mid + 1
		}
	}
	if nums[i] != target {
		return []int{-1, -1}
	}
	ret = append(ret, i)

	i, j = 0, len(nums) - 1
	//找到最右边出现的位置
	for i < j {
		mid := (i+j+1)/2
		if target >= nums[mid] {
			i = mid
		} else {
			j = mid - 1
		}
	}
	ret = append(ret, i)
	return ret
}

File: JsLeetcode/Binary-Tags/34. 在排序数组中查找元素的第一个和最后一个位置.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    if(nums.length == 0) {
        return [-1, -1];
    }
    var i = 0; 
    var j = nums.length - 1;
    while(i < j) {
        var mid = Math.floor((i + j) / 2);
        if(nums[mid] < target) {
            i = mid + 1;
        } else {
            j = mid;
        }
    };
    if(nums[i] != target) {
        return [-1, -1];
    }
    res = [];
    res.push(i);
    i, j = 0, nums.length - 1
    while(i < j) {
        mid = Math.floor((i+j+1)/2);
        if(nums[mid] > target) {
            j = mid -1
        } else {
            i = mid
        }
    }
};